### PR TITLE
[`DSA`] Fix vanillas SDPA with padding

### PR DESCRIPTION
**Pull request by vasqu**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48363) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48363)

As per title, this has been broken for DSA using SDPA for 2 core reasons
1. We always need a mask to account for causality in the indexer; SDPA can skip mask creation (`is_causal=True` kwarg)
2. SDPA creates a boolean mask but we assume additiveness -> properly account for bool based option

Fixes left padding issues when `index_topk < seq_len` and #48289 (cache vs no cache generation)

### CODE REVIEW

**Comment by vasqu** [comment]
run-slow: deepseek_v32, axk2, glm_moe_dsa

File: src/transformers/models/deepseek_v32/modular_deepseek_v32.py [hunk 1]
```diff
@@ -400,9 +400,20 @@ def forward(
...
 
+        # We need a float mask (additive bias) for DSA indexer
+        if causal_mask_mapping["deepseek_sparse_attention"].dtype == torch.bool:
+            min_dtype = torch.finfo(inputs_embeds.dtype).min
+            # we need 0s where the tokens should be taken into account, and -inf otherwise
+            causal_mask_mapping["deepseek_sparse_attention"] = torch.where(
+                causal_mask_mapping["deepseek_sparse_attention"],
+                torch.full((), 0.0, device=inputs_embeds.device, dtype=inputs_embeds.dtype),
+                min_dtype,
+            )
```
Comment (vasqu) [inline src/transformers/models/deepseek_v32/modular_deepseek_v32.py]: Not changing in the mask for now but was there any reason we kept the boolean mask for SDPA? Would be quite convenient to default to float for SDPA as well, e.g. here DSA, the position bias kwarg in sdpa attn path, and probably a few other models

cc @Cyrilvallez

Comment (ArthurZucker) [reply src/transformers/models/deepseek_v32/modular_deepseek_v32.py]: Main reason is memory: bool mask is 16 times cheaper than bf16 mask filled with inf. It's probably cheaper to fill it on the fly btw.

Comment (ArthurZucker) [reply src/transformers/models/deepseek_v32/modular_deepseek_v32.py]: also would be cleaner to only modify where its actually needed no?

Comment (vasqu) [reply src/transformers/models/deepseek_v32/modular_deepseek_v32.py]: Makes sense with the memory

With where it's needed do you mean the indexer? We could probably change the if else with the attn mask and use a masked fill (if bool) iiuc

File: src/transformers/models/deepseek_v32/modular_deepseek_v32.py [hunk 2]
```diff
@@ -208,8 +208,8 @@ def forward(
         hidden_states: torch.Tensor,
         q_resid: torch.Tensor,
         position_embeddings: tuple[torch.Tensor, torch.Tensor],
-        attention_mask: torch.Tensor | None,
-        position_ids: torch.Tensor,
+        attention_mask: torch.Tensor,
+        *,
```
Comment (vasqu) [inline src/transformers/models/deepseek_v32/modular_deepseek_v32.py]: Simplify by forcing mask creation otherwise we will need to add additional checks on whether we used sdpa or boolean

File: src/transformers/models/deepseek_v32/modular_deepseek_v32.py [hunk 3]
```diff
@@ -326,11 +328,11 @@ def forward(
                 .scatter(-1, topk_indices.long(), False)
                 .unsqueeze(1)
             )  # [B, 1, S, T]; True = masked
-            if attention_mask is None:
-                key_positions = torch.arange(key_states.shape[2], device=hidden_states.device)
-                index_mask = index_mask | (key_positions[None, None, None, :] > position_ids[:, None, :, None])
-                attention_mask = hidden_states.new_zeros((batch_size, 1, seq_length, key_states.shape[2]))
```
Comment (vasqu) [inline src/transformers/models/deepseek_v32/modular_deepseek_v32.py:332]: Similar, this assumes float mask and would require additional checks. Forcing the mask from the get go makes it much simpler (and we need a mask either way)

File: src/transformers/models/deepseek_v32/modular_deepseek_v32.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/deepseek_v32/modular_deepseek_v32.py]: Done, changed into properly checking between bool and float masks

**Comment by vasqu** [comment]
run-slow: deepseek_v32, axk2, glm_moe_dsa

File: src/transformers/masking_utils.py [hunk 4]
```diff
@@ -1511,7 +1512,8 @@ def create_recurrent_attention_mask(
     "compressed_sparse_attention": create_sliding_window_causal_mask,
     "heavily_compressed_attention": create_sliding_window_causal_mask,
     "minimax_m3_sparse": create_causal_mask,
-    "deepseek_sparse_attention": create_causal_mask,
+    # DSA always needs to materialize the mask to account for causality (no SDPA `is_cauasal` shortcut)
+    "deepseek_sparse_attention": partial(create_causal_mask, allow_is_causal_skip=False),
```
Comment (Cyrilvallez) [inline src/transformers/masking_utils.py:1516]: Oh interesting way to do here, we could use the same for qwen4_exp, and avoid overwriting the `create_mask_for_generate` there!
I can open a follow up PR

Comment (vasqu) [reply src/transformers/masking_utils.py:1516]: Oh yea, that's exactly what youd need as well :D I can also open another PR gimme a sec

Comment (vasqu) [reply src/transformers/masking_utils.py:1516]: #48456

File: src/transformers/models/deepseek_v32/modular_deepseek_v32.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/models/deepseek_v32/modular_deepseek_v32.py]: Yeah, my original reason was memory, though it's "only" 1x or 2x cheaper depending on dtype as bool are same as int8 in torch (also, I find it a bit easier to understand, like you know exactly what and what should not be taken into account, instead of taking the "opposite" for floats) - for dsa, in qwen4_exp it's also more practical as the indices you pick depend on valid positions, so you immediately know which ones are the valid positions

File: src/transformers/models/axk2/modeling_axk2.py [hunk 5]
```diff
@@ -803,6 +800,7 @@ def forward(
                 "attention_mask": attention_mask,
                 "past_key_values": past_key_values,
                 "position_ids": position_ids,
+                "allow_is_causal_skip": False,  # Always force creation to account for causality in the indexer
```
Comment (Cyrilvallez) [inline src/transformers/models/axk2/modeling_axk2.py:803]: Technically redundant with the partial, but nice to be explicit

**Review by Cyrilvallez — APPROVED** [review]
LGTM, thanks a lot!

**Review by ArthurZucker — APPROVED** [review]
nice!
